src/orderly/read.py

```diff
@@ -0,0 +1,65 @@
+import ast
+
+
+def orderly_read(path):
+    with open(path) as f:
+        src = ast.parse(f.read())
+    return _read_py(src)
+
+
+# In the R version of this function we do a more involved read, trying
+# to handle "static" versions of all of the orderly core
+# functions. However,we then never actually used that for anything, so
+# to avoid overcomplication we'll do the least possible here which is
+# just to read the parameters from the file, verify that it is only
+# called once, and that is called at the top-level.
+#
+# The return value does nod at future extension though.
+def _read_py(src):
+    ret = {"parameters": []}
+    for expr in src.body:
+        dat = _read_expr(expr)
+        if dat:
+            if dat["name"] == "parameters" and ret["parameters"]:
+                msg = f"Duplicate call to 'parameters()' on line {expr.lineno}"
+                raise Exception(msg)
+            ret["parameters"].append(dat["data"])
+    ret["parameters"] = ret["parameters"][0] if ret["parameters"] else {}
+    return ret
+
+
+def _read_expr(expr):
+    if isinstance(expr, ast.Assign):
+        return _read_expr(expr.value)
+    if _is_orderly_call(expr):
+        if expr.value.func.attr == "parameters":
+            return _read_parameters(expr.value)
+        return None
+    return None
+
+
+def _is_orderly_call(expr):
+    if not isinstance(expr, ast.Expr):
+        return False
+    if not isinstance(expr.value, ast.Call):
+        return False
+    call = expr.value
+    return call.func.value.id == "orderly" and call.func.attr
+
+
+def _read_parameters(call):
+    if call.args:
+        msg = "All arguments to 'parameters()' must be named"
+        raise Exception(msg)
+    data = {}
+    for kw in call.keywords:
+        nm = kw.arg
```

Comment:
Add a check for `kw.arg is None`. That can happen if someone passes in a `**kwargs`.

Reply:
I'm not sure I catch what you mean here. Do you have example failing test I can use?  Something like

```
_read_py(ast.parse("orderly.parameters(**kwargs)")
```

feels pretty unlikely no?

Reply:
thanks - at some point I'll go digging in the ast docs, but have added a check for this now which reports this case nicely at least